src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -368,11 +368,21 @@ public Task OnDataAsync(Http2Frame dataFrame, ReadOnlySequence<byte> payload)
                         {
                             RequestBodyPipe.Writer.Write(segment.Span);
                         }
-                        var flushTask = RequestBodyPipe.Writer.FlushAsync();
 
-                        // It shouldn't be possible for the RequestBodyPipe to fill up an return an incomplete task if
-                        // _inputFlowControl.Advance() didn't throw.
-                        Debug.Assert(flushTask.IsCompleted);
+                        // If the stream is completed go ahead and call RequestBodyPipe.Writer.Complete().
+                        // Data will still be available to the reader.
+                        if (endStream)
+                        {
+                            OnEndStreamReceived();
```

Comment:
Isn't OnEndStreamReceived already being called somewhere else?

Reply:
Nevermind, I see it bellow.

Reply:
Rather than calling OnEndStreamReceived in two places, change this to `if (!endStream)` for flush.